src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs

```diff
@@ -12,216 +13,174 @@ namespace System
+            // Do a quick search for the colon or uppercase letters
+            int index = str.AsSpan(start, end - start).LastIndexOfAny(s_asciiLetterUpperOrColonChars);
+            if (index >= 0)
             {
-                if (char.IsAsciiLetterUpper(str[i]))
+                Debug.Assert(!str.AsSpan(start, index - start).Contains(':'),
+                    "A colon should appear at most once, and must never be followed by letters.");
+
+                if (str[start + index] == ':')
                 {
-                    res = str.Substring(start, end - start).ToLowerInvariant();
-                    break;
+                    // Shrink the slice to only include chars before the colon
+                    end = start + index;
+
+                    // Look for uppercase letters again.
+                    // The index value doesn't matter anymore (nor does the search direction), just whether we've found anything
+                    index = str.AsSpan(start, index).IndexOfAnyInRange('A', 'Z');
                 }
-                if (str[i] == ':')
-                    end = i;
             }
 
-            res ??= str.Substring(start, end - start);
+            Debug.Assert(index == -1 || char.IsAsciiLetterUpper(str[start + index]));
+
+            if (index >= 0)
+            {
+                // We saw uppercase letters. Avoid allocating both the substring and the lower-cased variant.
+                return string.Create(end - start, (str, start), static (buffer, state) =>
+                {
+                    int newLength = state.str.AsSpan(state.start, buffer.Length).ToLowerInvariant(buffer);
+                    Debug.Assert(newLength == buffer.Length);
+                });
+            }
+
+            string res = str.Substring(start, end - start);
 
-            if (res == Localhost || res == Loopback)
+            if (res is Localhost or Loopback)
             {
                 loopback = true;
                 return Localhost;
             }
+
             return res;
         }
 
-        //
-        // IsValid
-        //
-        //  Determines whether a string is a valid domain name
-        //
-        //      subdomain -> <label> | <label> "." <subdomain>
-        //
-        // Inputs:
-        //  - name as Name to test
-        //  - starting position
-        //  - ending position
-        //
-        // Outputs:
-        //  The end position of a valid domain name string, the canonical flag if found so
-        //
-        // Returns:
-        //  bool
-        //
-        //  Remarks: Optimized for speed as a most common case,
-        //           MUST NOT be used unless all input indexes are verified and trusted.
-        //
-
-        internal static unsafe bool IsValid(char* name, int pos, ref int returnedEnd, ref bool notCanonical, bool notImplicitFile)
+        public static bool IsValid(ReadOnlySpan<char> hostname, bool iri, bool notImplicitFile, out int length)
```

Comment:
The `iri` parameter is always constant, so we could handle the differences with a generic, e.g.
```c#
private interface IIsValidImpl
{
    static abstract int IndexOfInvalidOrDelimiterChar(ReadOnlySpan<char> chars);
    static abstract bool CheckFirstLabelCharacter(char c);
    static abstract int DotIndex(ReadOnlySpan<char> chars);
    static abstract int CalculateLabelLength(ReadOnlySpan<char> label);
}
```
but it seemed like it was adding more code than it's worth (although arguably more readable?).